develop/tutorials/articles/210-application-security/04-authentication-pipelines/02-authentication-pipelines.markdown

```diff
@@ -0,0 +1,339 @@
+`UserLocalService` to look up the user's email address before checking it. This
+service is injected by the OSGi container because of the `@Reference`
+annotation. Note that the validator is also injected in this same manner, though
+it's configured not to fail if the implementation cannot be found. This allows
+this module to start regardless of its dependency on the validator
+implementation. In this case, this is safe because the error is handled by
+throwing an `AuthException` and logging the error. 
+
+Why would you want to do it this way? To error more gracefully. Because this is
+an `auth.pipeline.post` `Authenticator`, you presumably have other
+`Authenticator`s before this one checking credentials. If this one isn't
+working, you want to inform administrators with an error message rather than
+catastrophically failing and preventing users from logging in. 
+
+The only other Java code in this module is the Interface for the validator: 
+
+    package com.liferay.docs.emailaddressauthenticator.validator;
+
+    import aQute.bnd.annotation.ProviderType;
+
+    @ProviderType
+    public interface EmailAddressValidator {
+
+        public boolean isValidEmailAddress(String emailAddress);
+    }
+
+This defines a single method for checking the email address. 
+
+Next, you'll want to look at the validator module. 
+
+![Figure x: The validator project implements the Validator Interface and depends on the authenticator module. ](../../../images/auth-pipeline-validator-project.png)
+
+This module contains only one class. It implements the Validator interface: 
+
+    package com.liferay.docs.emailaddressvalidator.impl;
+
+    import java.util.Arrays;
+    import java.util.HashSet;
+    import java.util.Set;
+    import org.osgi.service.component.annotations.Component;
+    import com.liferay.docs.emailaddressauthenticator.validator.EmailAddressValidator;
+
+    @Component(
+        immediate = true,
+        property = {
+        },
+        service = EmailAddressValidator.class
+    )
+    public class EmailAddressValidatorImpl implements EmailAddressValidator {
+
+        @Override
+        public boolean isValidEmailAddress(String emailAddress) {
+            
+            if (_validEmailDomains.contains(
+                emailAddress.substring(emailAddress.indexOf('@')))) {
+            
+                return true;
+            }
+            return false;
+        }
+
+        private Set<String> _validEmailDomains = 
+            new HashSet<String>(Arrays.asList(new String[] {"@liferay.com", "@example.com"}));
+    }
+
+This code checks to make sure that the email address is from the @liferay.com or
+@example.com domains. The only other interesting part of this module is the
+Gradle build script, because it defines a compile-only dependency between the
+two projects. This is divided into two files: a `settings.gradle` and a
+`build.gradle`. 
+
+The `settings.gradle` file defines the location of the project (the
```

Comment:
I'm unfamiliar with this way of resolving projects, but I suspect this was contributed by someone more knowledgeable of this and it is needed to build outside of the Blade workspace?

Maybe since the settings.gradle contains quite a lot of stuff then there should be a hyperlink to another resource which explains it in more detail.

@sez11a